### ISSUE TITLE
Tracking Issue for `extended_varargs_abi_support`

### ISSUE THREAD

**Issue by Soveu** [rust-lang/rust#100189]
This is a tracking issue for extending available ABIs for var-args functions to support `aapcs`, `win64`, `sysv64` and `efiapi`.
The feature gate for the issue is `#![feature(extended_varargs_abi_support)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implementation - #97971
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR
    - ~~#116161~~ reverted by #136897

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Implementation history

- #97971
- #115860
- #116161 
- https://github.com/rust-lang/rust/pull/119587 note[^1]
- https://github.com/rust-lang/rust/pull/136897
- https://github.com/rust-lang/rust/pull/136948

### Unresolved Questions

- should `extern "system"` be included? the original feature only described itself as applying to `extern "efiapi"`, `extern "sysv64"`, and `extern "win64"`, and later the feature described itself as applying to `extern "aapcs"`, but the only case where `extern "system"` is meaningfully distinct is a case where it cannot be a vararg ABI. while MSVC sometimes translates `__stdcall` to be identical to the C ABI, we have previously rejected implementing MSVC's oddities:
  - https://github.com/rust-lang/rust/issues/87678

[^1]: this was not recorded as part of the impl history until after it was discovered this feature was initially stabilized without an FCP.

**Comment by Mark-Simulacrum** [comment]
Closing since this was stabilized and docs have been merged (https://github.com/rust-lang/reference/pull/1687).

**Comment by workingjubilee** [comment]
Reopening because no team ever FCPed this and there are legitimate reasons to question the precise implementation here. Accordingly I have opened https://github.com/rust-lang/rust/pull/136897

**Comment by workingjubilee** [comment]
Apparently part of the motive to fold `extern "system"` under this umbrella is https://github.com/rust-lang/rust/issues/110505 but the original feature was motivated by `extern "efiapi"`, so I would propose splitting out `extern "system"` into a new feature and giving it a separate review and decision.

**Comment by RalfJung** [comment]
As far as I can tell, the remaining functionality guarded by this feature is entirely uncontroversial, but someone needs to write a stabilization report to push it over the finish line.

**Comment by workingjubilee** [comment]
I had drafted a PR for stabilizing `extended_varargs_abi_support` but then noticed a lack of testing for it beyond the strict "feature gate" sense, i.e. I didn't see many tests that our support for linking against such functions and calling them would *work*. Only "static" does-it-compile tests, no "dynamic" tests.

**Comment by RalfJung** [comment]
Yeah I could imagine such tests being tricky... we'd have to build some C code and link against it.

**Comment by RalfJung** [comment]
I am proposing this for stabilization in https://github.com/rust-lang/rust/pull/144066.
